persistence/pre_persistence/app.go

```diff
@@ -0,0 +1,311 @@
+	app, exists, _ := m.GetApp(address)
+	if !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	cdc := Cdc()
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	// compute new values
+	stakedTokens, err := StringToBigInt(app.StakedTokens)
+	if err != nil {
+		return err
+	}
+	stakedTokensToAddI, err := StringToBigInt(amountToAdd)
+	if err != nil {
+		return err
+	}
+	stakedTokens.Add(stakedTokens, stakedTokensToAddI)
+	maxRelays, err := StringToBigInt(app.MaxRelays)
+	if err != nil {
+		return err
+	}
+	maxRelaysToAddI, err := StringToBigInt(maxRelaysToAdd)
+	if err != nil {
+		return err
+	}
+	maxRelays.Add(maxRelays, maxRelaysToAddI)
+	// update values
+	app.MaxRelays = BigIntToString(maxRelays)
+	app.StakedTokens = BigIntToString(stakedTokens)
+	app.Chains = chainsToUpdate
+	// marshal
+	bz, err := cdc.Marshal(app)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) DeleteApplication(address []byte) error {
+	if exists, _ := m.GetAppExists(address); !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	return db.Put(key, DeletedPrefixKey)
+}
+
+func (m *PrePersistenceContext) GetAppsReadyToUnstake(height int64, status int) (apps []*types.UnstakingActor, err error) { // TODO delete unstaking
+	db := m.Store()
+	unstakingKey := append(UnstakingAppPrefixKey, []byte(fmt.Sprintf("%d", height))...)
+	if has := db.Contains(unstakingKey); !has {
+		return nil, nil
+	}
+	val, err := db.Get(unstakingKey)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return make([]*types.UnstakingActor, 0), nil
+	}
+	unstakingApps := types.UnstakingActors{}
+	if err := proto.Unmarshal(val, &unstakingApps); err != nil {
+		return nil, err
+	}
+	for _, app := range unstakingApps.UnstakingActors {
+		apps = append(apps, app)
+	}
+	return
+}
+
+func (m *PrePersistenceContext) GetAppStatus(address []byte) (status int, err error) {
+	app, exists, err := m.GetApp(address)
+	if err != nil {
+		return ZeroInt, err
+	}
+	if !exists {
+		return ZeroInt, fmt.Errorf("does not exist in world state")
+	}
+	return int(app.Status), nil
+}
+
+func (m *PrePersistenceContext) SetAppUnstakingHeightAndStatus(address []byte, unstakingHeight int64, status int) error {
+	app, exists, err := m.GetApp(address)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	cdc := Cdc()
+	unstakingApps := types.UnstakingActors{}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	app.UnstakingHeight = unstakingHeight
+	app.Status = int32(status)
+	// marshal
+	bz, err := cdc.Marshal(app)
+	if err != nil {
+		return err
+	}
+	if err := db.Put(key, bz); err != nil {
+		return err
+	}
+	unstakingKey := append(UnstakingAppPrefixKey, []byte(fmt.Sprintf("%d", unstakingHeight))...)
+	if found := db.Contains(unstakingKey); found {
```

Comment:
I don't fully understand the business logic here...

We're trying to unstake an app above, but later find out that it is already being unstaked in the DB, so we are appending the unstaking actor? Shouldn't there only ever be exactly one unstaking actor corresponding to a specific address?

Lmk if we should take this offline.

Reply:
Let's take it offline

Reply:
Added it to our 1:1 notion doc.